xds/internal/xdsclient/authority.go

```diff
@@ -86,89 +98,132 @@ type authority struct {
+	// ordering specifies the order in which these channels are preferred for
+	// fallback.
+	xdsChannelConfigs []*xdsChannelWithConfig
+
+	// The current active xdsChannel. Here, active does not mean that the
+	// channel has a working connection to the server. It simply points to the
+	// channel that we are trying to work with, based on fallback logic.
+	activeXDSChannel *xdsChannelWithConfig
 }
 
 // authorityArgs is a convenience struct to wrap arguments required to create a
-// new authority. All fields here correspond directly to appropriate fields
-// stored in the authority struct.
+// new authority.
 type authorityArgs struct {
-	// The reason for passing server config and bootstrap config separately
-	// (although the former is part of the latter) is because authorities in the
-	// bootstrap config might contain an empty server config, and in this case,
-	// the top-level server config is to be used.
-	serverCfg          *bootstrap.ServerConfig
-	bootstrapCfg       *bootstrap.Config
-	serializer         *grpcsync.CallbackSerializer
-	resourceTypeGetter func(string) xdsresource.Type
-	watchExpiryTimeout time.Duration
-	backoff            func(int) time.Duration // Backoff for ADS and LRS stream failures.
-	logger             *grpclog.PrefixLogger
+	serverConfigs    bootstrap.ServerConfigs      // Server configs for the authority
+	name             string                       // Name of the authority
+	serializer       *grpcsync.CallbackSerializer // Callback serializer for invoking watch callbacks
+	getChannelForADS xdsChannelForADS             // Function to acquire a reference to an xdsChannel
+	logPrefix        string                       // Prefix for logging
 }
 
-func newAuthority(args authorityArgs) (*authority, error) {
+// newAuthority creates a new authority instance with the provided
+// configuration. The authority is responsible for managing the state of
+// resources requested from the management server, as well as acquiring and
+// releasing references to channels used to communicate with the management
+// server.
+//
+// Note that no channels to management servers are created at this time. Instead
+// a channel to the first server configuration is created when the first watch
+// is registered, and more channels are created as needed by the fallback logic.
+func newAuthority(args authorityArgs) *authority {
+	ctx, cancel := context.WithCancel(context.Background())
 	ret := &authority{
-		serverCfg:          args.serverCfg,
-		bootstrapCfg:       args.bootstrapCfg,
-		serializer:         args.serializer,
-		resourceTypeGetter: args.resourceTypeGetter,
-		watchExpiryTimeout: args.watchExpiryTimeout,
-		logger:             args.logger,
-		resources:          make(map[xdsresource.Type]map[string]*resourceState),
+		name:                      args.name,
+		watcherCallbackSerializer: args.serializer,
+		getChannelForADS:          args.getChannelForADS,
+		serializer:                grpcsync.NewCallbackSerializer(ctx),
```

Comment:
Why is the xDS Client serializer created inline but the watcherCallback is a knob? I see the watcher callback serializer getting passed from newClient as the client serializer? Is that to make all watches across all authorities serial?